wget/wget.go

```diff
@@ -0,0 +1,202 @@
+	}
+}
+
+func ParseURL(link string) error {
+	_, err := url.ParseRequestURI(link)
+	if err != nil {
+		return err
+	}
+	resp, err := http.Get(link)
+	if err != nil {
+		return fmt.Errorf("wget: error connecting url :%v", err)
+		//fmt.Fprintf(os.Stderr, "wget: error connecting url :%v\n", err)
+	}
+	defer resp.Body.Close()
+
+	bodySize, err := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	if err != nil {
+		return fmt.Errorf("wget: content length = %v :%v", bodySize, err)
+		//fmt.Fprintf(os.Stderr, "wget: error getting content length :%v\n", err)
+	}
+	return nil
+}
+
+func Execute() {
+	wg := &sync.WaitGroup{}
+	var linksList []string
+	dm := NewDownloadManager()
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+
+	// parsing urls
+	// if correct, adding them to slice
+	// else print error
+	for _, inputLink := range os.Args[1:] {
+		err := ParseURL(inputLink)
+		if err == nil {
+			linksList = append(linksList, inputLink)
+		} else {
+			fmt.Fprintf(os.Stderr, "wget: error parsing url: %v\n", err)
+		}
+	}
+
+	// locking map here
+	dm.fileListLock.Lock()
+	for _, link := range linksList {
+		dm.fileList[link] = &File{}
+		dm.fileListOrder = append(dm.fileListOrder, link)
+		wg.Add(1)
+		go dm.Download(link, wg)
+	}
+	dm.fileListLock.Unlock()
+
+	go dm.Print(ctx)
+
+	// waiting for all download goroutines
+	// to finish their work
+	wg.Wait()
+	// cancel ctx context which causes print func
+	// to return
+	cancel()
+	// ensure that last print will finish correctly
+	dm.WaitPrintEnd()
+	fmt.Println()
+}
+
+// Print uses context with select
+// and fileListOrder slice to prevent random
+// print order of map fields during iteration
+func (dm *DownloadManager) Print(ctx context.Context) {
+	defer close(dm.printFinishedCh)
+
+	print := func() {
+		fmt.Print("\r")
+		dm.fileListLock.Lock()
+		for _, key := range dm.fileListOrder {
+			if dm.fileList[key].totalSize != 0 {
+				fmt.Printf("%.2f %% ", float64(dm.fileList[key].counter)/float64(dm.fileList[key].totalSize)*100)
+			} else {
+				fmt.Print("0.00 %% ")
+			}
+		}
+		dm.fileListLock.Unlock()
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			print()
+			return
+		default:
+			print()
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+}
+
+// WaitPrintEnd helps us to be sure that
+// last call of Print function will be executed
+// before main gourutine ends
+func (dm *DownloadManager) WaitPrintEnd() {
+	<-dm.printFinishedCh
+	return
+}
+
+func (dm *DownloadManager) Download(link string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	resp, err := http.Get(link)
+	if err != nil {
+		//return false, 0, fmt.Errorf("wget: error connecting url :%v", err)
+		fmt.Fprintf(os.Stderr, "wget: error connecting url :%v\n", err)
+		return
+	}
+	defer resp.Body.Close()
+
+	bodySize, err := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	if err != nil {
+		//return false, 0, fmt.Errorf("wget: error getting content length :%v", err)
```

Comment:
dead code

Reply:
Have deleted dead code in commit 42bef00c07d5f5b947db74850f8533ee10ed0aa0